File: services/attendance-service/src/test/java/com/visor/school/attendance/integration/ClassLeaderAttendanceIntegrationTest.java
package com.visor.school.attendance.integration;

import com.visor.school.attendance.model.AttendanceRecord;
import com.visor.school.attendance.model.AttendanceSession;
import com.visor.school.attendance.model.AttendanceSessionStatus;
import com.visor.school.attendance.model.AttendanceStatus;
import com.visor.school.attendance.repository.AttendanceRepository;
import com.visor.school.attendance.repository.AttendanceSessionRepository;
import com.visor.school.attendance.service.AttendanceService;
import com.visor.school.attendance.service.AttendanceService.AttendanceEntry;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.amqp.rabbit.core.RabbitTemplate;

import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration test for class leader attendance collection and approval flow
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class ClassLeaderAttendanceIntegrationTest {

    @Autowired
    private AttendanceService attendanceService;

    @Autowired
    private AttendanceSessionRepository attendanceSessionRepository;

    @Autowired
    private AttendanceRepository attendanceRepository;

    @MockBean
    private RabbitTemplate rabbitTemplate;

    @Test
    void shouldCompleteClassLeaderAttendanceCollectionAndApprovalFlow() {
        // Given
        UUID classId = UUID.randomUUID();
        UUID teacherId = UUID.randomUUID();
        UUID classLeaderId = UUID.randomUUID();
        UUID studentId1 = UUID.randomUUID();
        UUID studentId2 = UUID.randomUUID();
        LocalDate date = LocalDate.now();

        // Step 1: Teacher creates session
        AttendanceSession session = attendanceService.createSession(
            classId,
            date,
            teacherId
        );

        assertEquals(AttendanceSessionStatus.PENDING, session.getStatus());
        assertNull(session.getDelegatedTo());

        // Step 2: Teacher delegates to class leader
        AttendanceSession delegatedSession = attendanceService.delegateToClassLeader(
            session.getId(),
            classLeaderId,
            classId
        );

        assertEquals(classLeaderId, delegatedSession.getDelegatedTo());
        assertEquals(AttendanceSessionStatus.PENDING, delegatedSession.getStatus());

        // Step 3: Class leader collects attendance
        List<AttendanceEntry> attendanceEntries = Arrays.asList(
            new AttendanceEntry(studentId1, AttendanceStatus.PRESENT, null),
            new AttendanceEntry(studentId2, AttendanceStatus.ABSENT, "Student absent")
        );

        AttendanceSession collectedSession = attendanceService.collectAttendanceByClassLeader(
            session.getId(),
            classLeaderId,
            attendanceEntries
        );

        assertEquals(AttendanceSessionStatus.COLLECTED, collectedSession.getStatus());
        assertNotNull(collectedSession.getCollectedAt());

        // Verify attendance records created
        List<AttendanceRecord> records = attendanceRepository.findBySessionId(session.getId());
        assertEquals(2, records.size());
        assertTrue(records.stream().anyMatch(it -> it.getStudentId().equals(studentId1) && it.getStatus() == AttendanceStatus.PRESENT));
        assertTrue(records.stream().anyMatch(it -> it.getStudentId().equals(studentId2) && it.getStatus() == AttendanceStatus.ABSENT));

        // Step 4: Teacher approves session
        AttendanceSession approvedSession = attendanceService.approveSession(session.getId(), teacherId);

        assertEquals(AttendanceSessionStatus.APPROVED, approvedSession.getStatus());
        assertEquals(teacherId, approvedSession.getApprovedBy());
        assertNotNull(approvedSession.getApprovedAt());

        // Verify records are linked to session
        List<AttendanceRecord> approvedRecords = attendanceRepository.findBySessionId(session.getId());
        assertTrue(approvedRecords.stream().allMatch(it -> it.getSessionId().equals(session.getId())));
    }

    @Test
    void shouldHandleSessionRejectionAndResubmission() {
        // Given
        UUID classId = UUID.randomUUID();
        UUID teacherId = UUID.randomUUID();
        UUID classLeaderId = UUID.randomUUID();
        UUID studentId = UUID.randomUUID();
        LocalDate date = LocalDate.now();

        // Step 1: Create and delegate session
        AttendanceSession session = attendanceService.createSession(classId, date, teacherId);
        attendanceService.delegateToClassLeader(session.getId(), classLeaderId, classId);

        // Step 2: Class leader collects
        AttendanceSession collectedSession = attendanceService.collectAttendanceByClassLeader(
            session.getId(),
            classLeaderId,
            List.of(new AttendanceEntry(studentId, AttendanceStatus.PRESENT, null))
        );

        assertEquals(AttendanceSessionStatus.COLLECTED, collectedSession.getStatus());

        // Step 3: Teacher rejects
        AttendanceSession rejectedSession = attendanceService.rejectSession(
            session.getId(),
            teacherId,
            "Incomplete attendance data"
        );

        assertEquals(AttendanceSessionStatus.REJECTED, rejectedSession.getStatus());
        assertEquals("Incomplete attendance data", rejectedSession.getRejectionReason());

        // Step 4: Class leader resubmits
        // We need to fetch from repository to get current state if we want to mimic entity reload, 
        // but 'rejectedSession' is detached/attached based on tx.
        // The service method saved it.
        
        AttendanceSession sessionEntity = attendanceSessionRepository.findById(session.getId()).orElseThrow();
        sessionEntity.resubmit();
        attendanceSessionRepository.save(sessionEntity);

        AttendanceSession resubmitted = attendanceSessionRepository.findById(session.getId()).orElseThrow();
        assertEquals(AttendanceSessionStatus.COLLECTED, resubmitted.getStatus());
        assertNull(resubmitted.getRejectionReason());
    }

    @Test
    void shouldPreventCollectionByNonDelegatedLeader() {
        // Given
        UUID classId = UUID.randomUUID();
        UUID teacherId = UUID.randomUUID();
        UUID classLeaderId = UUID.randomUUID();
        UUID otherStudentId = UUID.randomUUID();
        LocalDate date = LocalDate.now();

        AttendanceSession session = attendanceService.createSession(classId, date, teacherId);
        attendanceService.delegateToClassLeader(session.getId(), classLeaderId, classId);

        // When & Then
        assertThrows(IllegalArgumentException.class, () -> {
            attendanceService.collectAttendanceByClassLeader(
                session.getId(),
                otherStudentId, // Not the delegated leader
                List.of(new AttendanceEntry(UUID.randomUUID(), AttendanceStatus.PRESENT, null))
            );
        });
    }
}

File: services/attendance-service/src/test/kotlin/com/visor/school/attendanceservice/integration/ClassLeaderAttendanceIntegrationTest.kt
package com.visor.school.attendanceservice.integration

import com.visor.school.attendanceservice.model.AttendanceRecord
import com.visor.school.attendanceservice.model.AttendanceSession
import com.visor.school.attendanceservice.model.AttendanceSessionStatus
import com.visor.school.attendanceservice.model.AttendanceStatus
import com.visor.school.attendanceservice.repository.AttendanceRepository
import com.visor.school.attendanceservice.repository.AttendanceSessionRepository
import com.visor.school.attendanceservice.service.AttendanceService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID

/**
 * Integration test for class leader attendance collection and approval flow
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class ClassLeaderAttendanceIntegrationTest @Autowired constructor(
    private val attendanceService: AttendanceService,
    private val attendanceSessionRepository: AttendanceSessionRepository,
    private val attendanceRepository: AttendanceRepository
) {

    @Test
    fun `should complete class leader attendance collection and approval flow`() {
        // Given
        val classId = UUID.randomUUID()
        val teacherId = UUID.randomUUID()
        val classLeaderId = UUID.randomUUID()
        val studentId1 = UUID.randomUUID()
        val studentId2 = UUID.randomUUID()
        val date = LocalDate.now()

        // Step 1: Teacher creates session
        val session = attendanceService.createSession(
            classId = classId,
            date = date,
            createdBy = teacherId
        )

        assertEquals(AttendanceSessionStatus.PENDING, session.status)
        assertNull(session.delegatedTo)

        // Step 2: Teacher delegates to class leader
        val delegatedSession = attendanceService.delegateToClassLeader(
            sessionId = session.id,
            classLeaderId = classLeaderId,
            classId = classId
        )

        assertEquals(classLeaderId, delegatedSession.delegatedTo)
        assertEquals(AttendanceSessionStatus.PENDING, delegatedSession.status)

        // Step 3: Class leader collects attendance
        val attendanceEntries = listOf(
            AttendanceEntry(studentId1, AttendanceStatus.PRESENT, null),
            AttendanceEntry(studentId2, AttendanceStatus.ABSENT, "Student absent")
        )

        val collectedSession = attendanceService.collectAttendanceByClassLeader(
            sessionId = session.id,
            classLeaderId = classLeaderId,
            attendanceEntries = attendanceEntries
        )

        assertEquals(AttendanceSessionStatus.COLLECTED, collectedSession.status)
        assertNotNull(collectedSession.collectedAt)

        // Verify attendance records created
        val records = attendanceRepository.findBySessionId(session.id)
        assertEquals(2, records.size)
        assertTrue(records.any { it.studentId == studentId1 && it.status == AttendanceStatus.PRESENT })
        assertTrue(records.any { it.studentId == studentId2 && it.status == AttendanceStatus.ABSENT })

        // Step 4: Teacher approves session
        val approvedSession = attendanceService.approveSession(session.id, teacherId)

        assertEquals(AttendanceSessionStatus.APPROVED, approvedSession.status)
        assertEquals(teacherId, approvedSession.approvedBy)
        assertNotNull(approvedSession.approvedAt)

        // Verify records are linked to session
        val approvedRecords = attendanceRepository.findBySessionId(session.id)
        assertTrue(approvedRecords.all { it.sessionId == session.id })
    }

    @Test
    fun `should handle session rejection and resubmission`() {
        // Given
        val classId = UUID.randomUUID()
        val teacherId = UUID.randomUUID()
        val classLeaderId = UUID.randomUUID()
        val studentId = UUID.randomUUID()
        val date = LocalDate.now()

        // Step 1: Create and delegate session
        val session = attendanceService.createSession(classId, date, teacherId)
        attendanceService.delegateToClassLeader(session.id, classLeaderId, classId)

        // Step 2: Class leader collects
        val collectedSession = attendanceService.collectAttendanceByClassLeader(
            sessionId = session.id,
            classLeaderId = classLeaderId,
            attendanceEntries = listOf(AttendanceEntry(studentId, AttendanceStatus.PRESENT))
        )

        assertEquals(AttendanceSessionStatus.COLLECTED, collectedSession.status)

        // Step 3: Teacher rejects
        val rejectedSession = attendanceService.rejectSession(
            sessionId = session.id,
            teacherId = teacherId,
            reason = "Incomplete attendance data"
        )

        assertEquals(AttendanceSessionStatus.REJECTED, rejectedSession.status)
        assertEquals("Incomplete attendance data", rejectedSession.rejectionReason)

        // Step 4: Class leader resubmits
        val sessionEntity = attendanceSessionRepository.findById(session.id).get()
        sessionEntity.resubmit()
        attendanceSessionRepository.save(sessionEntity)

        val resubmitted = attendanceSessionRepository.findById(session.id).get()
        assertEquals(AttendanceSessionStatus.COLLECTED, resubmitted.status)
        assertNull(resubmitted.rejectionReason)
    }

    @Test
    fun `should prevent collection by non-delegated leader`() {
        // Given
        val classId = UUID.randomUUID()
        val teacherId = UUID.randomUUID()
        val classLeaderId = UUID.randomUUID()
        val otherStudentId = UUID.randomUUID()
        val date = LocalDate.now()

        val session = attendanceService.createSession(classId, date, teacherId)
        attendanceService.delegateToClassLeader(session.id, classLeaderId, classId)

        // When & Then
        assertThrows<IllegalArgumentException> {
            attendanceService.collectAttendanceByClassLeader(
                sessionId = session.id,
                classLeaderId = otherStudentId, // Not the delegated leader
                attendanceEntries = listOf(AttendanceEntry(UUID.randomUUID(), AttendanceStatus.PRESENT))
            )
        }
    }
}

// Data class for attendance entry
data class AttendanceEntry(
    val studentId: UUID,
    val status: AttendanceStatus,
    val notes: String? = null
)

